File: objectives.go
package okr2go

type Objective struct {
	Name        string      `json:"name"`
	Description string      `json:"description"`
	KeyResults  []KeyResult `json:"keyResults"`
}

func (o *Objective) FindKeyResult(resultID string) *KeyResult {
	for _, result := range o.KeyResults {
		if result.ID == resultID {
			return &result
		}
	}

	return nil
}

type KeyResult struct {
	ID           string   `json:"id"`
	Name         string   `json:"name"`
	Current      int64    `json:"current"`
	Target       int64    `json:"target"`
	Contributors []string `json:"contributors"`
	Comments     []string `json:"comments"`
}

File: ui/src/lib/objectives.ts
export async function listObjectives(): Promise<Objective[]> {
    const apiUrl = `/api/objectives`;

    return fetch(apiUrl)
        .then((res: Response) => res.json())
        .then((response: Objective[]) => {
            return response;
        });
}

export async function decreaseResultValue(objectiveId: number, keyResult: KeyResult): Promise<KeyResult> {
    const apiUrl = `/api/objectives/${objectiveId}/results/${keyResult.id}/minus`;

    return fetch(apiUrl)
        .then((res: Response) => res.json());
}

export async function increaseResultValue(objectiveId: number, keyResult: KeyResult): Promise<KeyResult> {
    const apiUrl = `/api/objectives/${objectiveId}/results/${keyResult.id}/plus`;

    return fetch(apiUrl)
        .then((res: Response) => res.json());
}


export async function createResult(objectiveId: number, keyResult: KeyResult): Promise<KeyResult> {
    const apiUrl = `/api/objectives/${objectiveId}/results`;

    return fetch(apiUrl, {
        method: 'POST',
        body: JSON.stringify(keyResult),
        headers: {
            'Content-Type': 'application/json'
        }
    })
        .then((res) => res.json())

}

export interface Objective {
    name: string
    description?: string
    keyResults?: KeyResult[]
}

export interface KeyResult {
    id: string,
    name: string
    current?: number,
    target: number,
    contributors?: string[],
    comments?: string[]
}
